dev_tools/notebooks/utils.py

```diff
@@ -104,20 +105,37 @@ def rewrite_notebook(notebook_path):
 
     used_patterns = set()
     with open(notebook_path, 'r') as original_file:
-        new_file_descriptor, new_file_path = tempfile.mkstemp(suffix='.ipynb')
-        with open(new_file_path, 'w') as new_file:
-            for line in original_file:
-                new_line = line
-                for pattern, replacement in patterns:
-                    new_line = pattern.sub(replacement, new_line)
-                    if new_line != line:
-                        used_patterns.add(pattern)
-                        break
-                new_file.write(new_line)
+        lines = original_file.readlines()
+    for i, line in enumerate(lines):
+        for pattern, replacement in patterns:
+            new_line = pattern.sub(replacement, line)
+            if new_line != line:
+                lines[i] = new_line
+                used_patterns.add(pattern)
+                break
 
     assert len(patterns) == len(used_patterns), (
         'Not all patterns where used. Patterns not used: '
         f'{set(x for x, _ in patterns) - used_patterns}'
     )
 
-    return new_file_descriptor, new_file_path
+    with tempfile.NamedTemporaryFile(mode='w', suffix='-rewrite.ipynb', delete=False) as new_file:
+        new_file.writelines(lines)
+
+    return new_file.name
+
+
+def remove_if_temporary_notebook(notebook_path: str):
+    """Delete temporary notebook if written by `rewrite_notebook`.
+
+    Do nothing if the specified notebook is not in the temporary directory
+    or if its filename does not end in '-rewrite.ipynb'
+
+    Use this to safely clean up notebooks created by `rewrite_notebook`.
+    """
+    if (
+        notebook_path.endswith('-rewrite.ipynb')
+        and os.path.isfile(notebook_path)
+        and os.path.dirname(notebook_path) == tempfile.gettempdir()
+    ):
```

Comment:
This implements the pattern wherein we make a decision (here: whether or not we need to rewrite a notebook into a temporary file) and then later try to work out what decision was taken to correlate earlier and later actions. This coding pattern is brittle since it's easy for the code to evolve in a way that breaks perfect correlation between the two decision points.

One way to solve this is to return a bool from `rewrite_notebook` indicating whether the file is temporary or not. This way we can ensure that the file is deleted if and only if it is temporary while also ensuring that the decision is made once.

If you really dislike two return values you could try to find the bool in the type by e.g. returning an instance of `tempfile.NamedTemporaryFile` or a file object (they both have `.name` attribute I think...). This is fine too as long as it can be made to work (haven't tried).

Reply:
Thanks for the feedback - per our offline talk, it is even simpler to have `rewrite_notebook` always create a temporary file which can be removed without any extra logic.

Done in 92477a7551605b05cdbeeefce0299e94a05a1508 and 0d5895f17110bcefe4303c572709a4cc6ae484f5.